### ISSUE TITLE
Sending a Non-Sized or Non-Send value on a channel, gives errors at non-obvious locations

### ISSUE THREAD

**Issue by hugoduncan** [rust-lang/rust#23950]
Expected: trying to send values of an incorrect type should be an error at the `send` call.

When compiling code that tries to send a values of a non-Sized type, an error is reported on the type's definition: [playpen](http://is.gd/6kdX1j)

``` rust
use std::error::Error;
use std::io;
use std::thread;
use std::sync::mpsc::channel;
use std::sync::Arc;

#[derive(Debug)]
struct A { // Error reported here
    b: Arc<Error>
}

fn main() {
    // Create a simple streaming channel
    let (tx, rx) = channel();
    let t = thread::scoped(move|| {
        tx.send(A{ b: Arc::new(io::Error::new(io::ErrorKind::Other, "other", None)) }).unwrap();
    });
    t.join();
    let v = rx.recv().unwrap();
    println!("{:?}",v);
}
```

When compiling code that tries to send a values of a non-Send type, an error is reported on the `channel()` call site: [playpen](http://is.gd/p85Gij)

``` rust
use std::boxed::Box;
use std::error::Error;
use std::io;
use std::thread;
use std::sync::mpsc::channel;

#[derive(Debug)]
struct A {
    b: Box<Error>
}

fn main() {
    // Create a simple streaming channel
    let (tx, rx) = channel(); // Error reported here
    let t = thread::scoped(move|| {
        tx.send(A{ b: Box::new(io::Error::new(io::ErrorKind::Other, "other", None)) }).unwrap();
    });
    t.join();
    let v = rx.recv().unwrap();
    println!("{:?}",v);
}
```

**Comment by apasel422** [comment]
Triage: Test cases need to be rewritten to account for the removal of `thread::scoped`.

**Comment by eefriedman** [comment]
```rust
use std::boxed::Box;
use std::error::Error;
use std::io;
use std::sync::mpsc::channel;

#[derive(Debug)]
struct A {
    b: Box<Error>
}

pub struct JoinGuard<'a, T: Send + 'a>(&'a T);
impl<'a, T: Send + 'a> JoinGuard<'a, T> {
    fn join(self) {}
}
pub fn scoped<'a, T, F>(_f: F) -> JoinGuard<'a, T> where T: Send + 'a, F: FnOnce() -> T, F: Send + 'a {
   unimplemented!()
}

fn main() {
    // Create a simple streaming channel
    let (tx, rx) = channel();
    let t = scoped(move|| {
        tx.send(A{ b: Box::new(io::Error::new(io::ErrorKind::Other, "other")) }).unwrap();
    });
    t.join();
    let v = rx.recv().unwrap();
    println!("{:?}",v);
}
```

Gives:

```
<anon>:22:13: 22:19 error: the trait `core::marker::Send` is not implemented for the type `std::error::Error + 'static` [E0277]
<anon>:22     let t = scoped(move|| {
                      ^~~~~~
<anon>:22:13: 22:19 help: see the detailed explanation for E0277
<anon>:22:13: 22:19 note: `std::error::Error + 'static` cannot be sent between threads safely
<anon>:22:13: 22:19 note: required because it appears within the type `Box<std::error::Error + 'static>`
<anon>:22:13: 22:19 note: required because it appears within the type `A`
<anon>:22:13: 22:19 note: required because it appears within the type `[closure@<anon>:22:20: 24:6 tx:std::sync::mpsc::Sender<A>]`
<anon>:22:13: 22:19 note: required by `scoped`
error: aborting due to previous error
playpen: application terminated with error code 101
```

Seems fine to me.

**Comment by Mark-Simulacrum** [comment]
I'm going to close, since the diagnostic appears to be fine today: it provides the type that isn't implementing `Send`. Please reopen if you disagree.

```
error[E0277]: the trait bound `std::error::Error + 'static: std::marker::Send` is not satisfied
  --> ./test.rs:22:13
   |
22 |     let t = scoped(move|| {
   |             ^^^^^^ the trait `std::marker::Send` is not implemented for `std::error::Error + 'static`
   |
   = note: `std::error::Error + 'static` cannot be sent between threads safely
   = note: required because of the requirements on the impl of `std::marker::Send` for `std::ptr::Unique<std::error::Error + 'static>`
   = note: required because it appears within the type `std::boxed::Box<std::error::Error + 'static>`
   = note: required because it appears within the type `A`
   = note: required because of the requirements on the impl of `std::marker::Send` for `std::sync::mpsc::Sender<A>`
   = note: required because it appears within the type `[closure@./test.rs:22:20: 24:6 tx:std::sync::mpsc::Sender<A>]`
   = note: required by `scoped`

error: aborting due to previous error
```
